### PR TITLE
`sage.schemes`: Reformat doctests, add `# optional` annotations

### PR DESCRIPTION
**Pull request by mkoeppe**
### 📚 Description

Adding doctest tags `# optional - sage.rings.finite_rings`, `...number_field`, `...padics`.

While going through the doctests line by line, I also made the following changes:
- some coding style fixes in the doctests (such as adding spaces around some operators and after commas, following PEP 8) - improved the readability of them in the HTML format by breaking long lines to avoid having to scroll horizontally
- improved indentation of input and output, in particular when morphisms are displayed
- made better use of the horizontal space in the doctests of some modules in `sage.schemes.toric`, which were formatted for a very narrow layout
- improved the sphinx markup of some docstrings. 

The doctest tags are preparation for being able to test parts of `sage.schemes` in a modularized setting, in which not all rings are available. See https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#doctest-only-dependencies

Part of:
- #29705 

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### ⌛ Dependencies

- Depends on #35230 (not merged here) for the definition of the feature `sage.rings.finite_rings`.

The mass edits adding `# optional` tags were made using the following Emacs macro.
```elisp
(defun sage-copy-optional-annotation ()
  "In a 'sage: ' line of a docstring, copy '# optional' from a previous line and
advance to the end of the next 'sage: ' line or to the end of the current docstring.
If invoked elsewhere, just advance to the end of the next 'sage: ' line."
  (interactive)
  (if (save-excursion (beginning-of-line)
		      (looking-at " *sage:"))
      (let ((tab-stop-list
	     (list (save-excursion
		     (previous-line)
		     (end-of-line)
		     (re-search-backward "# *optional.*$")
		     (- (match-beginning 0)
			(save-excursion
			  (beginning-of-line) (point)))))))
	(end-of-line)
	(just-one-space)
	(tab-to-tab-stop)
	(insert (match-string-no-properties 0))
	(re-search-forward "sage: \\|\"\"\"")
	(end-of-line))
    (re-search-forward "sage:")
    (end-of-line)))

(with-eval-after-load "python-mode"
  (define-key python-mode-map (kbd "C-M-;") 'sage-copy-optional-annotation))
```

### CODE REVIEW

File: src/sage/schemes/affine/affine_homset.py [hunk 1]
```diff
@@ -210,14 +210,14 @@ def points(self, **kwds):
...
-            sage: A2(F).points()
+            sage: F = GF(3)                                                             # optional - sage.rings.finite_rings
+            sage: A2(F).points()                                                        # optional - sage.rings.finite_rings
             [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
 
         ::
 
             sage: A.<x,y> = ZZ[]
-            sage: I = A.ideal(x^2-y^2-1)
+            sage: I = A.ideal(x^2 - y^2-1)
```
Comment (kwankyu) [inline src/sage/schemes/affine/affine_homset.py]: Needs spaces around the second `-` too.

File: src/sage/schemes/affine/affine_morphism.py [hunk 2]
```diff
@@ -355,7 +347,7 @@ def _fastpolys(self):
 
             sage: P.<x,y> = AffineSpace(QQ, 2)
             sage: H = Hom(P, P)
-            sage: f = H([x^2+y^2, y^2/(1+x)])
+            sage: f = H([x^2 + y^2, y^2/(1+x)])
```
Comment (kwankyu) [inline src/sage/schemes/affine/affine_morphism.py:350]: needs spaces around the second `+`.

File: src/sage/schemes/affine/affine_morphism.py [hunk 3]
```diff
@@ -464,58 +456,56 @@ def homogenize(self, n):
 
             sage: A.<x,y> = AffineSpace(CC, 2)
             sage: H = Hom(A, A)
-            sage: f = H([(x^2-2)/(x*y), y^2-x])
+            sage: f = H([(x^2-2)/(x*y), y^2 - x])
```
Comment (kwankyu) [inline src/sage/schemes/affine/affine_morphism.py:459]: needs spaces

File: src/sage/schemes/affine/affine_morphism.py [hunk 4]
```diff
@@ -464,58 +456,56 @@ def homogenize(self, n):
...
+              Defn: Defined on coordinates by sending (x0 : x1 : x2) to
+                    (9*x1^2 : 3*x1*x2 : x2^2)
 
         ::
 
             sage: R.<t> = PolynomialRing(ZZ)
             sage: A.<x,y> = AffineSpace(R, 2)
             sage: H = Hom(A, A)
-            sage: f = H([(x^2-2)/y, y^2-x])
+            sage: f = H([(x^2-2)/y, y^2 - x])
```
Comment (kwankyu) [inline src/sage/schemes/affine/affine_morphism.py:483]: needs spaces

File: src/sage/schemes/affine/affine_morphism.py (hunk 2 above)
Comment (mkoeppe) [reply src/sage/schemes/affine/affine_morphism.py:350]: I'm not adding spaces within nested sums

File: src/sage/schemes/affine/affine_homset.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/schemes/affine/affine_homset.py]: Done in a38133637b

File: src/sage/schemes/elliptic_curves/constructor.py [hunk 5]
```diff
@@ -146,7 +146,8 @@ class EllipticCurveFactory(UniqueFactory):
 
         sage: E = EllipticCurve(CC, [0,0,1,-1,0])
         sage: E
-        Elliptic Curve defined by y^2 + 1.00000000000000*y = x^3 + (-1.00000000000000)*x over Complex Field with 53 bits of precision
+        Elliptic Curve defined by y^2 + 1.00000000000000*y = x^3 + (-1.00000000000000)*x
+        over Complex Field with 53 bits of precision
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/constructor.py]: Sometimes I see
```
Elliptic Curve ...
 over ...
```
Other times
```
Elliptic Curves ...
over ...
```
Need to be consistent. Here and elsewhere.

File: src/sage/schemes/elliptic_curves/ell_number_field.py [hunk 6]
```diff
@@ -19,8 +19,8 @@
 
 EXAMPLES::
 
-    sage: K.<i> = NumberField(x^2+1)
-    sage: E = EllipticCurve([0,4+i])
+    sage: K.<i> = NumberField(x^2 + 1)
+    sage: E = EllipticCurve([0, 4+i])
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/ell_number_field.py:23]: No `optional` tag in this file?

File: src/sage/schemes/elliptic_curves/height.py [hunk 7]
```diff
@@ -778,14 +780,17 @@ def __init__(self, E):
...
             sage: E = EllipticCurve([0,0,0,0,1])
             sage: EllipticCurveCanonicalHeight(E)
-            EllipticCurveCanonicalHeight object associated to Elliptic Curve defined by y^2 = x^3 + 1 over Rational Field
+            EllipticCurveCanonicalHeight object associated to
+            Elliptic Curve defined by y^2 = x^3 + 1 over Rational Field
 
         An example over a number field::
 
-            sage: K.<i>=QuadraticField(-1)
+            sage: K.<i> = QuadraticField(-1)
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/height.py]: No `optional` tag in this file?

File: src/sage/schemes/elliptic_curves/hom_composite.py [hunk 8]
```diff
@@ -23,10 +23,11 @@
     2^143
     sage: EllipticCurveHom_composite(E, P)
     Composite morphism of degree 11150372599265311570767859136324180752990208 = 2^143:
-      From: Elliptic Curve defined by y^2 = x^3 + x over Finite Field in z2 of size 33451117797795934712303577408972542258970623^2
+      From: Elliptic Curve defined by y^2 = x^3 + x
+            over Finite Field in z2 of size 33451117797795934712303577408972542258970623^2
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/hom_composite.py:27]: No `optional` tag in this file?

File: src/sage/schemes/elliptic_curves/constructor.py (hunk 5 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/constructor.py]: I am standardizing on the first format now

File: src/sage/schemes/elliptic_curves/ell_number_field.py (hunk 6 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/ell_number_field.py:23]: Added in 9fcebc3c80.

File: src/sage/schemes/elliptic_curves/height.py (hunk 7 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/height.py]: Added in 9fcebc3c80

File: src/sage/schemes/elliptic_curves/hom_composite.py (hunk 8 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/hom_composite.py:27]: Added in 9fcebc3c80

File: src/sage/schemes/berkovich/berkovich_space.py [hunk 9]
```diff
@@ -315,59 +314,59 @@ class Berkovich_Cp_Affine(Berkovich_Cp):
...
         Affine Berkovich line over Cp(3) of precision 1000
 
     Points of Berkovich space can be created from points of
     extensions of `\QQ_p`::
 
-        sage: B = Berkovich_Cp_Affine(3)
-        sage: A.<a> = Qp(3).extension(x^3 - 3)
-        sage: B(a)
+        sage: B = Berkovich_Cp_Affine(3)                                                # optional - sage.rings.padics
+        sage: A.<a> = Qp(3).extension(x^3 - 3)                                          # optional - sage.rings.number_field
```
Comment (kwankyu) [inline src/sage/schemes/berkovich/berkovich_space.py]: Extensions of p-adic fields seems to still belong to `sage.rings.padics`.

Comment (mkoeppe) [reply src/sage/schemes/berkovich/berkovich_space.py]: Thanks, fixed in ed39a64

File: src/sage/schemes/elliptic_curves/hom_frobenius.py [hunk 10]
```diff
@@ -55,7 +55,11 @@
     sage: E = EllipticCurve([GF(17^6).gen(), 0])
     sage: pi = EllipticCurveHom_frobenius(E)
     sage: pihat = pi.dual(); pihat
-    Isogeny of degree 17 from Elliptic Curve defined by y^2 = x^3 + (15*z6^5+5*z6^4+8*z6^3+12*z6^2+11*z6+7)*x over Finite Field in z6 of size 17^6 to Elliptic Curve defined by y^2 = x^3 + z6*x over Finite Field in z6 of size 17^6
+    Isogeny of degree 17
+     from Elliptic Curve defined by y^2 = x^3 + (15*z6^5+5*z6^4+8*z6^3+12*z6^2+11*z6+7)*x
+          over Finite Field in z6 of size 17^6
+       to Elliptic Curve defined by y^2 = x^3 + z6*x
+          over Finite Field in z6 of size 17^6
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/hom_frobenius.py:67]: No `optional` tags in this file?

File: src/sage/schemes/elliptic_curves/isogeny_class.py [hunk 11]
```diff
@@ -362,7 +362,9 @@ def isogenies(self, fill=False):
 
             sage: isocls = EllipticCurve('15a3').isogeny_class()
             sage: f = isocls.isogenies()[0][1]; f
-            Isogeny of degree 2 from Elliptic Curve defined by y^2 + x*y + y = x^3 + x^2 - 5*x + 2 over Rational Field to Elliptic Curve defined by y^2 + x*y + y = x^3 + x^2 - 80*x + 242 over Rational Field
+            Isogeny of degree 2
+              from Elliptic Curve defined by y^2 + x*y + y = x^3 + x^2 - 5*x + 2 over Rational Field
+                to Elliptic Curve defined by y^2 + x*y + y = x^3 + x^2 - 80*x + 242 over Rational Field
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/isogeny_class.py:367]: No `optional` tags in this file?

File: src/sage/schemes/elliptic_curves/isogeny_small_degree.py [hunk 12]
```diff
@@ -796,13 +824,34 @@ def isogenies_5_0(E, minimal_models=True):
...
+          from Elliptic Curve defined by y^2 = x^3 + 10 over Finite Field in a of size 13^2
+            to Elliptic Curve defined by y^2 = x^3 + (3*a+12)*x + (11*a+12) over Finite Field in a of size 13^2,
+         Isogeny of degree 5
+          from Elliptic Curve defined by y^2 = x^3 + 10 over Finite Field in a of size 13^2
+            to Elliptic Curve defined by y^2 = x^3 + (a+4)*x + (11*a+12) over Finite Field in a of size 13^2,
+         Isogeny of degree 5
+          from Elliptic Curve defined by y^2 = x^3 + 10 over Finite Field in a of size 13^2
+            to Elliptic Curve defined by y^2 = x^3 + (9*a+10)*x + (11*a+12) over Finite Field in a of size 13^2]
+
+        sage: K.<a> = NumberField(x**6 - 320*x**3 - 320)
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/isogeny_small_degree.py]: No `optional` tags in this file?

Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/isogeny_small_degree.py]: Fixed in 289d256

File: src/sage/schemes/elliptic_curves/isogeny_class.py (hunk 11 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/isogeny_class.py:367]: Fixed in d5d51b5

File: src/sage/schemes/elliptic_curves/hom_frobenius.py (hunk 10 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/hom_frobenius.py:67]: Fixed in 5698dfa

File: src/sage/schemes/overview.py [hunk 13]
```diff
@@ -117,9 +117,9 @@
   ::
 
         sage: R.<x> = ZZ[]
-        sage: S.<t> = R.quo(x^2+5)
-        sage: P.<X,Y,Z> = ProjectiveSpace(2, S)
-        sage: P(S)
+        sage: S.<t> = R.quo(x^2 + 5)                                                    # optional - sage.rings.number_field
+        sage: P.<X,Y,Z> = ProjectiveSpace(2, S)                                         # optional - sage.rings.number_field
+        sage: P(S)                                                                      # optional - sage.rings.number_field
```
Comment (kwankyu) [inline src/sage/schemes/overview.py]: `S` is not really a number field, but a quotient of a polynomial ring.
```
sage: R.<x> = ZZ[]
sage: S.<t> = R.quo(x^2+5)
sage: type(S)
<class 'sage.rings.polynomial.polynomial_quotient_ring.PolynomialQuotientRing_domain_with_category'>
```

File: src/sage/schemes/toric/library.py [hunk 14]
```diff
@@ -862,25 +790,16 @@ def Cube_face_fan(self, names='z+', base_ring=QQ):
...
-            N( 1,  1,  1),
-            N( 1, -1,  1),
-            N(-1,  1,  1),
-            N(-1, -1,  1),
-            N(-1, -1, -1),
-            N(-1,  1, -1),
-            N( 1, -1, -1),
-            N( 1,  1, -1)
+            N( 1,  1,  1),    N( 1, -1,  1),    N(-1,  1,  1),    N(-1, -1,  1),
+            N(-1, -1, -1),    N(-1,  1, -1),    N( 1, -1, -1),    N( 1,  1, -1)
```
Comment (kwankyu) [inline src/sage/schemes/toric/library.py:802]: This is not necessarily an improvement, as the data in the rectangular format is harder to read than in the vertical format.

But it is okay. I am not suggesting to revert.

**Comment by kwankyu** [comment]
I looked through the changes. They look good to me. 

This is a lot of work, and a great improvement of the quality of docstrings in the schemes module.

File: src/sage/schemes/toric/library.py (hunk 14 above)
Comment (mkoeppe) [reply src/sage/schemes/toric/library.py:802]: I could change these from the 2x4 format to a 4x2 format as a compromise?

File: src/sage/schemes/overview.py (hunk 13 above)
Comment (mkoeppe) [reply src/sage/schemes/overview.py]: Indeed, fixed in 5065211aa9

File: src/sage/schemes/toric/library.py (hunk 14 above)
Comment (kwankyu) [reply src/sage/schemes/toric/library.py:802]: No. Just leave them as they are. I am okay with it.

File: src/sage/schemes/elliptic_curves/constructor.py (hunk 5 above)
Comment (kwankyu) [reply src/sage/schemes/elliptic_curves/constructor.py]: Okay for this PR.

But for standardizing how to split a long doctest output and formatting nicely multiline output, I don't know what is right. A related idea is to respect the actual output in the sage terminal. I don't know whether we should or not. This issue perhaps belongs to https://doc.sagemath.org/html/en/developer/coding_basics.html#fine-points-on-styles

**Review by kwankyu — APPROVED** [review]
Thanks.

File: src/sage/schemes/elliptic_curves/constructor.py (hunk 5 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/constructor.py]: I don't think there's a need for setting a policy for developers; to me, anything in the spectrum between the raw actual output (assisted by `sage -fixdoctests`) and the delicate doctest poetry that I have attempted here is fine. I don't think there is a downside to reformatting the output with the goal of readability / user experience. There is potential for mild disappointment when users try out an example and the output does not look as pretty as it appears in the documentation. But even then the structure-indicating reformatted output guides the user through parsing the complicated raw output. And Sage has various ways to adjust output, using `%display latex`, `%display ascii_art` etc.; the difference between documentation and reality may inspire users to explore these facilities (or even improve our `_repr_` code).

**Comment by mkoeppe** [comment]
Thanks very much for reviewing this mega-patch!

**Comment by mkoeppe** [comment]
Easy rebase.

**Comment by vbraun** [comment]
```
sage -t --long --warn-long 37.7 --random-seed=123 src/sage/schemes/elliptic_curves/ell_generic.py
**********************************************************************
File "src/sage/schemes/elliptic_curves/ell_generic.py", line 1394, in sage.schemes.elliptic_curves.ell_generic.EllipticCurve_generic.change_ring
Failed example:
    h = F2.hom([a[0]], F4)                                                # optional - sage.rings.finite_rings
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/rings/finite_rings/homset.py", line 100, in __call__
        return FiniteFieldHomomorphism_generic(self, im_gens, base_map=base_map, check=check)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/rings/finite_rings/hom_finite_field.pyx", line 239, in sage.rings.finite_rings.hom_finite_field.FiniteFieldHomomorphism_generic.__init__
        RingHomomorphism_im_gens.__init__(self, parent, im_gens, base_map=base_map, check=check)
      File "sage/rings/morphism.pyx", line 1832, in sage.rings.morphism.RingHomomorphism_im_gens.__init__
        raise ValueError("relations do not all (canonically) map to 0 under map determined by images of generators")
    ValueError: relations do not all (canonically) map to 0 under map determined by images of generators

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/rings/finite_rings/homset.py", line 109, in __call__
        return self._coerce_impl(im_gens)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/release/Sage/src/sage/rings/finite_rings/homset.py", line 132, in _coerce_impl
        raise TypeError
    TypeError

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.schemes.elliptic_curves.ell_generic.EllipticCurve_generic.change_ring[3]>", line 1, in <module>
        h = F2.hom([a[Integer(0)]], F4)                                                # optional - sage.rings.finite_rings
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/structure/parent_gens.pyx", line 300, in sage.structure.parent_gens.ParentWithGens.hom
        return parent.Parent.hom(self, im_gens, codomain, base_map=base_map, category=category, check=check)
      File "sage/structure/parent.pyx", line 1427, in sage.structure.parent.Parent.hom
        return self.Hom(codomain, **Hom_kwds)(im_gens, **kwds)
      File "/home/release/Sage/src/sage/rings/finite_rings/homset.py", line 111, in __call__
        raise TypeError("images do not define a valid homomorphism")
    TypeError: images do not define a valid homomorphism
**********************************************************************
File "src/sage/schemes/elliptic_curves/ell_generic.py", line 1398, in sage.schemes.elliptic_curves.ell_generic.EllipticCurve_generic.change_ring
Failed example:
    E.change_ring(h)                                                      # optional - sage.rings.finite_rings
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.schemes.elliptic_curves.ell_generic.EllipticCurve_generic.change_ring[5]>", line 1, in <module>
        E.change_ring(h)                                                      # optional - sage.rings.finite_rings
                      ^
    NameError: name 'h' is not defined
**********************************************************************
1 item had failures:
   2 of   7 in sage.schemes.elliptic_curves.ell_generic.EllipticCurve_generic.change_ring
    [554 tests, 2 failures, 3.71 s]
----------------------------------------------------------------------
sage -t --long --warn-long 37.7 --random-seed=123 src/sage/schemes/elliptic_curves/ell_generic.py  # 2 doctests failed
----------------------------------------------------------------------

```

**Pull request merged**